### PR TITLE
Add deformable conv v2 op,test=develop

### PR DESCRIPTION
**Pull request by baiyfbupt**
Test on K40m GPU, CUDA 9.0, CUDNN 7.1
deformable_conv and conv2d both on `num_filters=32, filter_size=3, stride=1, padding=1, dilation=1` configuration

| Event                | Calls | Total   | CPU Time (Ratio)        | GPU Time (Ratio)       | Min.     | Max.     | Ave.      |
| -------------------- | ----- | ------- | ----------------------- | ---------------------- | -------- | -------- | --------- |
| conv2d               | 80    | 21805.5 | 16456.982404 (0.754716) | 5348.544296 (0.245284) | 0.168048 | 5500.73  | 272.569   |
| conv2d_grad          | 80    | 102.503 | 15.649915(0.152677)     | 86.853337 (0.847323)   | 1.22332  | 1.52524  | 1.28129   |
| deformable_conv      | 80    | 5.46764 | 5.320669 (0.973119)     | 0.146974 (0.026881)    | 0.043983 | 0.345927 | 0.0683455 |
| deformable_conv_grad | 80    | 16.2551 | 14.926100 (0.918241)    | 1.329006 (0.081759)    | 0.139445 | 1.77225  | 0.203189  |

test=develop

### CODE REVIEW

File: paddle/fluid/operators/modulated_deformable_conv_op.cc [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+                      "divided by deformable groups.");
+
+    if (in_dims[0] > im2col_step) {
+      PADDLE_ENFORCE_EQ(
+          in_dims[0] % im2col_step, 0U,
+          "Input batchsize must be smaller than or divide im2col_step");
+    }
+
+    for (size_t i = 0; i < strides.size(); ++i) {
+      PADDLE_ENFORCE_GT(strides[i], 0U, "incorrect stride size");
```
Comment (heavengate) [inline paddle/fluid/operators/modulated_deformable_conv_op.cc]: 错误信息里面把 i 也输出吧，可用性好一些，下同

File: paddle/fluid/operators/modulated_deformable_conv_op.cc [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        .SetDefault(1);
+    AddAttr<int>("im2col_step",
+                 "im2col maximum number of image per computation")
+        .SetDefault(64);
+    AddComment(R"DOC(
+**Modulated Deformable Convolution Operator**
+
+Compute 2-D deformable convolution on 4-D input.
+
+The deformable convolution operation is described in https://arxiv.org/abs/1703.06211
```
Comment (heavengate) [inline paddle/fluid/operators/modulated_deformable_conv_op.cc]: 文档可以详细一些，给出数学公式

**Comment by baiyfbupt** [comment]
test=develop

File: paddle/fluid/operators/modulated_deformable_conv_op.cc (hunk 2 above)
Comment (baiyfbupt) [reply paddle/fluid/operators/modulated_deformable_conv_op.cc]: Done, thanks

File: paddle/fluid/operators/modulated_deformable_conv_op.cc (hunk 1 above)
Comment (baiyfbupt) [reply paddle/fluid/operators/modulated_deformable_conv_op.cc]: Done

**Comment by baiyfbupt** [comment]
deformable_api

**Comment by baiyfbupt** [comment]
@shanyi15 Please take a look at this api change. Thanks

File: paddle/fluid/operators/deformable_conv_op.cc [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    std::vector<int> strides = ctx->Attrs().Get<std::vector<int>>("strides");
+    std::vector<int> paddings = ctx->Attrs().Get<std::vector<int>>("paddings");
+    std::vector<int> dilations =
+        ctx->Attrs().Get<std::vector<int>>("dilations");
+    int groups = ctx->Attrs().Get<int>("groups");
+    int deformable_groups = ctx->Attrs().Get<int>("deformable_groups");
+    int im2col_step = ctx->Attrs().Get<int>("im2col_step");
+
+    PADDLE_ENFORCE(in_dims.size() == 4,
+                   "Conv intput should be 4-D tensor, get %u", in_dims.size());
```
Comment (heavengate) [inline paddle/fluid/operators/deformable_conv_op.cc]: intput -> input

File: paddle/fluid/operators/deformable_conv_op.cu [hunk 4]
```diff
@@ -0,0 +1,745 @@
...
+static constexpr int kNumCUDAThreads = 512;
+static constexpr int kNumMaximumNumBlocks = 4096;
+
+static inline int NumBlocks(const int N) {
+  return std::min((N + kNumCUDAThreads - 1) / kNumCUDAThreads,
+                  kNumMaximumNumBlocks);
+}
+
+template <typename T>
+__device__ T dmcn_get_gradient_weight(T argmax_h, T argmax_w, const int h,
```
Comment (qingqing01) [inline paddle/fluid/operators/deformable_conv_op.cu]: dmcn_get_gradient_weight -> DmcnGetGradientWeight

File: paddle/fluid/operators/deformable_conv_op.cu [hunk 5]
```diff
@@ -0,0 +1,745 @@
...
+    weight = (h + 1 - argmax_h) * (argmax_w + 1 - w);
+  if (h == argmax_h_high && w == argmax_w_low)
+    weight = (argmax_h + 1 - h) * (w + 1 - argmax_w);
+  if (h == argmax_h_high && w == argmax_w_high)
+    weight = (argmax_h + 1 - h) * (argmax_w + 1 - w);
+  return weight;
+}
+
+template <typename T>
+__global__ void modulated_deformable_col2im_gpu_kernel(
```
Comment (qingqing01) [inline paddle/fluid/operators/deformable_conv_op.cu]: 注意函数命名风格，下同

File: paddle/fluid/operators/deformable_conv_op.cu [hunk 6]
```diff
@@ -0,0 +1,745 @@
...
+  int index = blockIdx.x * blockDim.x + threadIdx.x;
+  int offset = blockDim.x * gridDim.x;
+  for (size_t thread = index; thread < nthreads; thread += offset) {
+    const int j = (thread / width_col / height_col / batch_size) % kernel_w;
+    const int i =
+        (thread / width_col / height_col / batch_size / kernel_w) % kernel_h;
+    const int c =
+        thread / width_col / height_col / batch_size / kernel_w / kernel_h;
+
+    const int deformable_group_index = c / channel_per_deformable_group;
```
Comment (qingqing01) [inline paddle/fluid/operators/deformable_conv_op.cu:78]: 回头测下和conv的时间对比，这块的除法可以作为后续的优化

**Comment by baiyfbupt** [comment]
> Event	Calls	Total	CPU Time (Ratio)	GPU Time (Ratio)	Min.	Max.	Ave.
> conv2d	80	21805.5	16456.982404 (0.754716)	5348.544296 (0.245284)	0.168048	5500.73	272.569
> conv2d_grad	80	102.503	15.649915(0.152677)	86.853337 (0.847323)	1.22332	1.52524	1.28129
> deformable_conv	80	5.46764	5.320669 (0.973119)	0.146974 (0.026881)	0.043983	0.345927	0.0683455
> deformable_conv_grad	80	16.2551	14.926100 (0.918241)	1.329006 (0.081759)	0.139445	1.77225	0.203189
> test=develop

File: paddle/fluid/operators/deformable_conv_op.cu (hunk 4 above)
Comment (baiyfbupt) [reply paddle/fluid/operators/deformable_conv_op.cu]: Done

**Review by qingqing01 — APPROVED** [review]

**Review by xsrobin — APPROVED** [review]
LGTM

**Review by shanyi15 — APPROVED** [review]

**Pull request merged**